packages/mobile/src/screens/chat-screen/ChatScreen.tsx

```diff
@@ -92,22 +101,83 @@ export const ChatScreen = () => {
 
   const { params } = useRoute<'Chat'>()
   const { chatId } = params
-  const url = `/chat/${encodeUrlName(chatId)}`
-  const [iconOpacity, setIconOpacity] = useState(0.5)
-  const [text, setText] = useState('')
+  const url = `/chat/${encodeUrlName(chatId ?? '')}`
+  const [iconOpacity, setIconOpacity] = useState(ICON_BLUR)
+  const [value, setValue] = useState('')
 
+  const userId = encodeHashId(useSelector(getUserId))
+  const chat = useSelector((state) => getChat(state, chatId ?? ''))
+  const [otherUser] = useSelector((state) => getOtherChatUsers(state, chatId))
   const chatMessages = useSelector((state) =>
     getChatMessages(state, chatId ?? '')
   )
   const status = useSelector((state) =>
     getChatMessagesStatus(state, chatId ?? '')
   )
+  const summary = useSelector((state) =>
+    getChatMessagesSummary(state, chatId ?? '')
+  )
+  const flatListRef = useRef<RNFlatList>(null)
 
+  // A ref so that the unread separator doesn't disappear immediately when the chat is marked as read
+  // Using a ref instead of state here to prevent unwanted flickers.
+  // The chat/chatId selectors will trigger the rerenders necessary.
+  const chatFrozenRef = useRef(chat)
   useEffect(() => {
-    dispatch(fetchMoreMessages({ chatId }))
-  }, [dispatch, chatId])
+    if (chat && chatId !== chatFrozenRef.current?.chat_id) {
+      // Update the unread indicator when chatId changes
+      chatFrozenRef.current = chat
+    }
+  }, [chat, chatId])
 
-  const [otherUser] = useSelector((state) => getOtherChatUsers(state, chatId))
+  useEffect(() => {
+    if (chatId && status === Status.IDLE) {
+      // Initial fetch
+      dispatch(fetchMoreMessages({ chatId }))
+      dispatch(setActiveChat({ chatId }))
+    }
+  }, [dispatch, chatId, status, summary])
+
+  const earliestUnreadIndex = useMemo(() => {
```

Comment:
Possible perf thing to flag, this is going to run whenever we add a new message and will iterate over the entire messages array, and then a second time with the indexOf. We could probably use findIndex to cut down one of these calls, which would also stop iterating over the array once the condition is met. Do we need to even compute this if we're in an active chat, though?

Reply:
Good call!

Reply:
Actually unfortunately I think we do need to recalc every time chatMessages changes, bc if we send a message or a new message comes in, the unread tag will move if earliestUnreadIndex doesn't update.